src/storage/v2/vertex_info_cache.cpp

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "storage/v2/vertex_info_cache.hpp"
+
+#include "storage/v2/property_value.hpp"
+
+namespace memgraph::storage {
+
+VertexInfoCache::VertexInfoCache() = default;
+VertexInfoCache::~VertexInfoCache() = default;
+VertexInfoCache::VertexInfoCache(VertexInfoCache &&) noexcept = default;
+VertexInfoCache &VertexInfoCache::operator=(VertexInfoCache &&) noexcept = default;
+
+auto VertexInfoCache::getExists(View view, Vertex const *vertex) const -> std::optional<bool> {
+  auto const &cache = (view == View::OLD) ? old_.existsCache_ : new_.existsCache_;
+  if (cache.empty()) return std::nullopt;
+
+  auto const it = cache.find(vertex);
+  if (it == cache.end()) return std::nullopt;
+  return {it->second};
+}
+void VertexInfoCache::storeExists(View view, Vertex const *vertex, bool res) {
+  auto &cache = (view == View::OLD) ? old_.existsCache_ : new_.existsCache_;
+  cache.emplace(vertex, res);
+}
+auto VertexInfoCache::getDeleted(View view, Vertex const *vertex) const -> std::optional<bool> {
+  auto const &cache = (view == View::OLD) ? old_.deletedCache_ : new_.deletedCache_;
+  if (cache.empty()) return std::nullopt;
+
+  auto const it = cache.find(vertex);
+  if (it == cache.end()) return std::nullopt;
+  return {it->second};
+}
+void VertexInfoCache::storeDeleted(View view, Vertex const *vertex, bool res) {
+  auto &cache = (view == View::OLD) ? old_.deletedCache_ : new_.deletedCache_;
+  cache.emplace(vertex, res);
+}
+
+void VertexInfoCache::invalidate(Vertex const *vertex) {
+  new_.existsCache_.erase(vertex);
+  new_.deletedCache_.erase(vertex);
+  // Should we invalidate other caches proactively, should be guarded by existence/deletion?
```

Comment:
I would invalidate everything.

Reply:
Yep testing proved that it needed to invalidate. Code has been changed.